cpp/src/copying/scatter.cu

```diff
@@ -177,6 +179,58 @@ void scalar_scatter(const std::vector<gdf_scalar>& source,
 
 }
 
+inline bool validate_scatter_map(gdf_column const& scatter_map,
+                          cudf::table const& input_table) {
+  CUDF_EXPECTS(scatter_map.dtype == GDF_INT32,
+      "scatter_map is not GDF_INT32 column.");
+  CUDF_EXPECTS(not cudf::has_nulls(scatter_map),
+      "Scatter map cannot contain null elements.");
+  CUDF_EXPECTS(scatter_map.size == input_table.num_rows(),
+      "scatter_map length is not equal to number of rows in input table.");
+  if (scatter_map.size == 0 ||
+      input_table.num_columns() == 0 ||
+      input_table.num_rows() == 0)
+    return false;
+  return true;
+}
+
+template<typename InputIterator>
+std::vector<cudf::table>
+ordered_scatter_to_tables(cudf::table const& input_table,
```

Comment:
> You also don't need to reserve, you can use the vector size constructor. :)

This won't have the same result.  

```
    std::vector<cudf::table> output_tables(num_groups);
    for (gdf_size_type group = 0; group < num_groups; group++) {
        output_tables.push_back(
```

The result will be `output_table.size() == 2 * num_groups)`. 

Constructing with `num_groups` will make the initial size be `num_groups`, and then doing `push_back` for another `num_groups` makes the final size two times `num_groups`. 

`reserve` allocates the necessary space without actually changing the `size()` of the vector.

Reply:
> You also don't need to `reserve`, you can use the vector size constructor. :)

If I use vector size constructor, it will call `cudf::table()` constructor `num_groups` times. since each `cudf::table` object is constructed inside `copy_if`, I don't need this.
I avoid this by creating empty vector and calling `reserve`.

Reply:
> ordered_scatter_to_tables is only called from one place. It takes two iterators, even though the first one is always a wrapper around zero. There's no reason to use an iterator here since all you do is a `for` loop over the range of the iterators, and then you immediately dereference the iterator to get a sequential integer out of it.

This `ordered_scatter_to_tables` will be used for `cudf::groups_to_tables`,  There, iterators will be useful. (`cudf::groups_to_tables` moved out of this PR).

Reply:
But do they really need to be iterators if the thing you are iterating on is an int?

Reply:
The idea was to pass unique groupid as scatter_map and iterate on it. 
It is possible to implement  `groups_to_tables`  with just an int. I will update it to int.